src/web/mvc/Program.cs

```diff
@@ -1,26 +1,121 @@
+
+builder.Services.AddPiranha(options =>
 {
-    public class Program
+    /*
+     * This will enable automatic reload of .cshtml
+     * without restarting the application. However since
+     * this adds a slight overhead it should not be
+     * enabled in production.
+     */
+    options.AddRazorRuntimeCompilation = true;
+
+    options.UseCms();
+    options.UseManager();
+
+#if (UseBlobStorage)
+    options.UseBlobStorage(builder.Configuration.GetConnectionString("blobstorage"), naming: Piranha.Azure.BlobStorageNaming.UniqueFolderNames);
+#else
+    options.UseFileStorage(naming: Piranha.Local.FileStorageNaming.UniqueFolderNames);
+#endif
+    options.UseImageSharp();
+    options.UseTinyMCE();
+    options.UseMemoryCache();
+
+    var connectionString = builder.Configuration.GetConnectionString("piranha");
+#if (UseSQLServer)
+    options.UseEF<SQLServerDb>(db => db.UseSqlServer(connectionString));
+    options.UseIdentityWithSeed<IdentitySQLServerDb>(db => db.UseSqlServer(connectionString));
+#elif (UseMySql)
+    options.UseEF<MySqlDb>(db => db.UseMySql(connectionString, ServerVersion.AutoDetect(connectionString)));
+    options.UseIdentityWithSeed<IdentityMySQLDb>(db => db.UseMySql(connectionString, ServerVersion.AutoDetect(connectionString)));
+#elif (UsePostgreSql)
+    options.UseEF<PostgreSqlDb>(db => db.UseNpgsql(connectionString));
+    options.UseIdentityWithSeed<IdentityPostgreSQLDb>(db => db.UseNpgsql(connectionString));
+#else
+    options.UseEF<SQLiteDb>(db => db.UseSqlite(connectionString));
+    options.UseIdentityWithSeed<IdentitySQLiteDb>(db => db.UseSqlite(connectionString));
+#endif
+
+    /*
+     * Here you can configure the different permissions
+     * that you want to use for securing content in the
+     * application.
+    options.UseSecurity(o =>
     {
-        public static void Main(string[] args)
-        {
-            CreateHostBuilder(args).Build().Run();
-        }
-
-        public static IHostBuilder CreateHostBuilder(string[] args) =>
-            Host.CreateDefaultBuilder(args)
-                .ConfigureWebHostDefaults(webBuilder =>
-                {
-                    webBuilder.UseStartup<Startup>();
-                });
-    }
+        o.UsePermission("WebUser", "Web User");
+    });
+     */
+
+    /*
+     * Here you can specify the login url for the front end
+     * application. This does not affect the login url of
+     * the manager interface.
+    options.LoginUrl = "login";
+     */
+});
+
+var app = builder.Build();
+
+// For application behind a reverse proxy, accept forwarded headers
+// Nginx: https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/linux-nginx?view=aspnetcore-6.0
+// Apache: https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/linux-apache?view=aspnetcore-6.0
+//app.UseForwardedHeaders(new ForwardedHeadersOptions
+//{
+//    ForwardedHeaders = ForwardedHeaders.All
+//});
+
+var api = app.Services.GetService<IApi>();

```

Comment:
@goranhell @Darth-Jurassic The error you're getting is because `IApi` is a scoped service, and just like the error message says a scoped service can't be resolved from the root provider, it needs a scope to be resolved. To go around this issue, line 88 should be changed to:

~~~
var api = app.Services.CreateScope().ServiceProvider.GetService<IApi>();
~~~

Reply:
![image](https://user-images.githubusercontent.com/4937730/148733801-d7f365b6-00af-40a1-b484-2e9ec8d4a2ee.png)
![image](https://user-images.githubusercontent.com/4937730/148733830-67571e40-5bed-4ddf-a98b-587c3bf91e60.png)
Nope, it works pretty well. And it worked the same in Startup. Startup isn't scoped either, so this behavior is pretty similar, but without this unnatural Startup lifecycle.

Reply:
This is actually not true. When injecting a service into the `Startup.Configure()` method ASP.NET creates a scope before calling the method which is why there's no issue injecting scoped services in this context. I can reproduce the issues @goranhell is getting by converting the example projects in the main repository to use your startup code. It fails with the error message:

~~~
Unhandled exception. System.InvalidOperationException: Cannot resolve scoped service 'Piranha.IApi' from root provider.
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteValidator.ValidateResolution(Type serviceType, IServiceScope scope, IServiceScope rootScope)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetService[T](IServiceProvider provider)
   at Program.<Main>$(String[] args) in C:\Proj\github\piranha.core\examples\RazorWeb\Program.cs:line 34
~~~

Where line `34` in this `Program.cs` is:

~~~
var api = app.Services.GetService<IApi>();
~~~

Reply:
For reference: https://khalidabuhakmeh.com/resolve-services-in-aspnet-core-startup

Reply:
My line 34 from this template with SQLite is options.UseIdentityWithSeed<IdentitySQLiteDb>(db => db.UseSqlite(connectionString)); and start of a commented section. Please attach full Program.cs in your comments because otherwise I have to guess what it is.

I have attached the browser and debugger screenshots earlier, it works, it just works. Maybe the problem is in partial refactoring or using ASP .NET 5 as runtime (same human hands problem that I had when migrating my solution to ASP .NET 6). But your template creates ASP .NET 6 project, and the Program.cs DI is new in ASP .NET 6, and I thought it's OK to use ASP .NET 6 approach and to use ASP .NET 6 features since it is released and it is LTS.

My Program.cs just from MVC template with SQLite is:

```
using System.Reflection;
//using Microsoft.AspNetCore.HttpOverrides;
using Microsoft.EntityFrameworkCore;
using Piranha;
using Piranha.AttributeBuilder;
using Piranha.AspNetCore.Identity.SQLite;
using Piranha.Data.EF.SQLite;
using Piranha.Manager.Editor;

var builder = WebApplication.CreateBuilder(args);

builder.Services.AddPiranha(options =>
{
    /*
     * This will enable automatic reload of .cshtml
     * without restarting the application. However since
     * this adds a slight overhead it should not be
     * enabled in production.
     */
    options.AddRazorRuntimeCompilation = true;

    options.UseCms();
    options.UseManager();

    options.UseFileStorage(naming: Piranha.Local.FileStorageNaming.UniqueFolderNames);
    options.UseImageSharp();
    options.UseTinyMCE();
    options.UseMemoryCache();

    var connectionString = builder.Configuration.GetConnectionString("piranha");
    options.UseEF<SQLiteDb>(db => db.UseSqlite(connectionString));
    options.UseIdentityWithSeed<IdentitySQLiteDb>(db => db.UseSqlite(connectionString));

    /*
     * Here you can configure the different permissions
     * that you want to use for securing content in the
     * application.
    options.UseSecurity(o =>
    {
        o.UsePermission("WebUser", "Web User");
    });
     */

    /*
     * Here you can specify the login url for the front end
     * application. This does not affect the login url of
     * the manager interface.
    options.LoginUrl = "login";
     */
});

var app = builder.Build();

// For application behind a reverse proxy, accept forwarded headers
// Nginx: https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/linux-nginx?view=aspnetcore-6.0
// Apache: https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/linux-apache?view=aspnetcore-6.0
//app.UseForwardedHeaders(new ForwardedHeadersOptions
//{
//    ForwardedHeaders = ForwardedHeaders.All
//});

if (app.Environment.IsDevelopment())
{
    app.UseDeveloperExceptionPage();
}
else
{
    app.UseExceptionHandler("/Error");
    // The default HSTS value is 30 days. You may want to change this for production scenarios, see https://aka.ms/aspnetcore-hsts.
    app.UseHsts();
}

var api = app.Services.GetService<IApi>();

// Initialize Piranha
App.Init(api);

// Build content types
new ContentTypeBuilder(api)
    .AddAssembly(Assembly.GetExecutingAssembly())
    .Build()
    .DeleteOrphans();

// Configure Tiny MCE
EditorConfig.FromFile("editorconfig.json");

// Middleware setup
app.UsePiranha(options =>
{
    options.UseManager();
    options.UseTinyMCE();
    options.UseIdentity();
});

await app.RunAsync();
```

https://docs.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection?view=aspnetcore-6.0 — see the «Services injected into Program.cs» — looks like it is a new feature in ASP.NET 6 as there is no such a section in 5.0 version.

There is still an opportunity to limit Program.cs injections lifetime with scopes, but it's optional and to make it right I must be sure (can you confirm it will be so?) Piranha.App.Init and Piranha.AttributeBuilder.ContentTypeBuilder uses only values from IApi, but never binds to IApi (and I see that ContentTypeBuilder does, it is not IDisposable and an ObjectDisposedException never occured only because IApi was never called in ContentTypeBuilder finalizer, but it is potentially the case). Currently the scoped version can look like the following:

```
await using (var scope = app.Services.CreateAsyncScope())
{
    var api = scope.ServiceProvider.GetRequiredService<IApi>();
    // Initialize Piranha
    App.Init(api);
    
    // Build content types
    await (await new ContentTypeBuilder(api)
        .AddAssembly(Assembly.GetExecutingAssembly())
        .BuildAsync())
        .DeleteOrphansAsync();
}
```

But to guarantee it doesn't throw an ObjectDisposedException the ContentTypeBuilder must be IDisposable to have a place for scoped destruction implementation. Otherwise current Program.cs call is fully equivalent to a scope that disposes after app.RunAsync is complete giving no use and no need of explicit scope definition.

Now about the InvalidOperationException...

Your exception tells that your app behaves according to https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/web-host?view=aspnetcore-6.0#scope-validation , are you still using WebHost.CreateDefaultBuilder instead of WebApplication.CreateBuilder ? Note that the PR templates use WebApplication.CreateBuilder, and that's why it allows to work with injections in Program.cs.